crates/outbound-networking/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+use anyhow::Context;
+use url::Url;
+
+/// Try to parse the url that may or not include the provided scheme.
+///
+/// If the parsing fails, the url is appended with the scheme and parsing
+/// is tried again.
+pub fn parse_url_with_host(url: &str, scheme: &str) -> anyhow::Result<Url> {
+    match Url::parse(url) {
+        Ok(url) if url.has_host() => Ok(url),
+        first_try => {
+            let second_try = format!("{scheme}://{url}")
+                .as_str()
+                .try_into()
+                .context("could not convert into a url");
+            match (second_try, first_try.map_err(|e| e.into())) {
+                (Ok(u), _) => Ok(u),
+                // Return an error preferring the error from the first attempt if present
+                (_, Err(e)) | (Err(e), _) => Err(e),
+            }
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum AllowedHosts {
+    All,
+    SpecificHosts(Vec<AllowedHost>),
+}
+
+impl AllowedHosts {
+    pub fn parse<S: AsRef<str>>(hosts: &[S]) -> anyhow::Result<AllowedHosts> {
+        // TODO: do we support this?
+        // if hosts.len() == 1 && hosts[0].as_ref() == "insecure:allow-all" {
+        //     return Ok(Self::All);
+        // }
+        let mut allowed = Vec::with_capacity(hosts.len());
+        for host in hosts {
+            allowed.push(AllowedHost::parse(host)?)
+        }
+        Ok(Self::SpecificHosts(allowed))
+    }
+
+    pub fn allows<U: TryInto<Url>>(&self, url: U) -> bool {
+        match self {
+            AllowedHosts::All => true,
+            AllowedHosts::SpecificHosts(hosts) => {
+                let Ok(url) = url.try_into() else {
+                    return false;
+                };
+                hosts.iter().any(|h| h.allows(&url))
+            }
+        }
+    }
+}
+
+impl Default for AllowedHosts {
+    fn default() -> Self {
+        Self::SpecificHosts(Vec::new())
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct AllowedHost {
+    scheme: Option<String>,
+    host: String,
+    port: u16,
+}
+
+impl AllowedHost {
+    fn parse<U: AsRef<str>>(url: U) -> anyhow::Result<Self> {
+        let url_str = url.as_ref();
+        let url: anyhow::Result<Url> = url_str
+            .try_into()
+            .with_context(|| format!("could not convert {url_str:?} into a url"));
```

Comment:
Leaving this comment once for all the places: URL should be capitalised in user-facing messages and I will suffer a mild toe injury on this hill.